### PR TITLE
Orca Spark Estimator Interface

### PR DESCRIPTION
**Pull request by cyita**

### CODE REVIEW

**Comment by jason-dai** [comment]
Do we need to make sure these estimators all have the same default batch size?

File: pyzoo/zoo/orca/learn/pytorch/estimator.py [hunk 1]
```diff
@@ -271,7 +271,8 @@ def predict(self, data, batch_size=4):
         predicted_rdd = self.model.predict(data_rdd, batch_size=batch_size)
         return convert_predict_to_xshard(predicted_rdd)
 
-    def evaluate(self, data, validation_methods=None, batch_size=32):
+    def evaluate(self, data, batch_size=32, feature_cols=None, labels_cols=None,
+                 validation_methods=None):
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/pytorch/estimator.py:275]: change to `validation_metrics`

File: pyzoo/zoo/orca/learn/pytorch/estimator.py [hunk 2]
```diff
@@ -218,8 +218,8 @@ def __init__(self, model, loss, optimizer, model_dir=None, bigdl_type="float"):
         self.model = TorchModel.from_pytorch(model)
         self.estimator = SparkEstimator(self.model, optimizer, model_dir, bigdl_type=bigdl_type)
 
-    def fit(self, data, epochs=1, batch_size=32, validation_data=None, validation_methods=None,
-            checkpoint_trigger=None):
+    def fit(self, data, epochs=1, batch_size=32, feature_cols=None, labels_cols=None,
+            validation_data=None, validation_methods=None, checkpoint_trigger=None):
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/pytorch/estimator.py:222]: change to `validation_metrics`

File: pyzoo/zoo/orca/learn/bigdl/estimator.py [hunk 3]
```diff
@@ -171,7 +173,8 @@ def predict(self, data, batch_size=8, feature_cols="features", sample_preprocess
             raise ValueError("Data should be XShards or Spark DataFrame, but get " +
                              data.__class__.__name__)
 
-    def evaluate(self, data, validation_methods=None, batch_size=32):
+    def evaluate(self, data, batch_size=32, feature_cols=None, labels_cols=None,
+                 validation_methods=None):
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/bigdl/estimator.py:177]: change to `validation_metrics`

File: pyzoo/zoo/orca/learn/bigdl/estimator.py [hunk 4]
```diff
@@ -128,21 +130,21 @@ def fit(self, data, epochs, feature_cols="features", labels_cols="label", batch_
...
                 else:
-                    assert isinstance(val_data, SparkXShards), "val_data should be a XShards"
-                    val_feature_set = FeatureSet.sample_rdd(val_data.rdd.flatMap(to_sample))
+                    assert isinstance(validation_data, SparkXShards), "val_data should be a XShards"
+                    val_feature_set = FeatureSet.sample_rdd(validation_data.rdd.flatMap(to_sample))
                 if self.log_dir is not None and self.app_name is not None:
                     self.estimator.set_tensorboard(self.log_dir, self.app_name)
                 self.estimator.train(train_feature_set, self.loss, end_trigger, checkpoint_trigger,
-                                     val_feature_set, val_methods, batch_size)
+                                     val_feature_set, validation_methods, batch_size)
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/bigdl/estimator.py:147]: change to `validation_metrics`

File: pyzoo/zoo/orca/learn/bigdl/estimator.py [hunk 5]
```diff
@@ -81,36 +81,38 @@ def __init__(self, *, model, loss, optimizer=None, feature_preprocessing=None,
         self.app_name = None
         self.is_nnframe_fit = False
 
-    def fit(self, data, epochs, feature_cols="features", labels_cols="label", batch_size=32,
-            caching_sample=True, val_data=None, val_trigger=None, val_methods=None,
-            checkpoint_trigger=None):
+    def fit(self, data, epochs, batch_size=32, feature_cols="features", labels_cols="label",
+            caching_sample=True, validation_data=None, validation_trigger=None,
+            validation_methods=None, checkpoint_trigger=None):
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/bigdl/estimator.py:86]: change to `validation_metrics`

File: pyzoo/zoo/orca/learn/bigdl/estimator.py [hunk 6]
```diff
@@ -81,36 +81,38 @@ def __init__(self, *, model, loss, optimizer=None, feature_preprocessing=None,
...
                 .setCachingSample(caching_sample).setFeaturesCol(feature_cols)\
                 .setLabelCol(labels_cols)
 
-            if val_data is not None:
-                assert isinstance(val_data, DataFrame), "val_data should be a spark DataFrame."
-                assert val_trigger is not None and val_methods is not None, \
+            if validation_data is not None:
+                assert isinstance(validation_data, DataFrame), \
+                    "val_data should be a spark DataFrame."
+                assert validation_trigger is not None and validation_methods is not None, \
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/bigdl/estimator.py:110]: change to `validation_metrics`

File: pyzoo/zoo/orca/learn/bigdl/estimator.py [hunk 7]
```diff
@@ -81,36 +81,38 @@ def __init__(self, *, model, loss, optimizer=None, feature_preprocessing=None,
...
                     "You should provide val_trigger and val_methods if you provide val_data."
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/bigdl/estimator.py:111]: "validation_trigger" and ""validation_metrics

File: pyzoo/zoo/examples/orca/learn/pytorch/mnist/lenet_mnist.py [hunk 8]
```diff
@@ -103,8 +103,8 @@ def main():
     adam = torch.optim.Adam(model.parameters(), args.lr)
     est = Estimator.from_torch(model=model, optimizer=adam, loss=criterion)
     est.fit(data=train_loader, epochs=args.epochs, validation_data=test_loader,
-            validation_methods=[Accuracy()], checkpoint_trigger=EveryEpoch())
-    result = est.evaluate(data=test_loader, validation_methods=[Accuracy()])
+            validation_metrics=[Accuracy()], checkpoint_trigger=EveryEpoch())
+    result = est.evaluate(data=test_loader, validation_metrics=[Accuracy()])
```
Comment (jason-dai) [inline pyzoo/zoo/examples/orca/learn/pytorch/mnist/lenet_mnist.py:107]: need to update https://github.com/intel-analytics/analytics-zoo/blob/master/docs/docs/Orca/orca-pytorch-quickstart.md

File: pyzoo/zoo/orca/learn/base_estimator.py [hunk 9]
```diff
@@ -39,5 +39,29 @@ def save(self, model_path):
...
+    @abstractmethod
+    def get_train_summary(self, tag=None):
+        raise NotImplementedError
+
+    @abstractmethod
+    def get_validation_summary(self, tag=None):
+        raise NotImplementedError
+
+    @abstractmethod
+    def set_tensorboard(self, log_dir, app_name):
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/base_estimator.py:66]: Should these go to SparkEstimator?

**Comment by jason-dai** [comment]
Change the class names as follows:
- `TFOptimizerWrapper` --> `TFSparkEstimator`
- `TFKerasWrapper` --> `KerasSparkEstimator`
- `PytorchSparkEstimatorWrapper` --> `PyTorchSparkEstimator`
- `BigDLEstimatorWrapper` --> `BigDLEstimator`

File: pyzoo/zoo/orca/learn/base_estimator.py (hunk 9 above)
Comment (cyita) [reply pyzoo/zoo/orca/learn/base_estimator.py]: Yes. Ray estimators are hard to support these methods now.

**Comment by jason-dai** [comment]
LGTM

**Comment by cyita** [comment]
Jenkins passed http://10.239.47.210:18888/job/ZOO-PR-Validation/4880/

**Pull request merged**